### ISSUE TITLE
Pod traffic not taking advantage of interfaces with multiple tx queues on AWS

### ISSUE THREAD

**Issue by lbernail** [cilium/cilium#18311]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

We run Cilium on AWS in `ipam=eni` mode and we isolate pod interfaces from the main node one (using `"first-interface-index": 1`). At high throughput, we noticed a significant amount of TCP retransmits that we tracked back to qdic drops (by looking at the output of `tc -s qdisc show dev ens6`). When we looked more closely we noticed that:
- we have one qdisc per tx queue on the ENI (expected)
- all traffic from pods use the first qdisc and the first tx queue of the ENI (unexpected and this limits the maximum throughput we can achieve)

We performed a few tests to reproduce and isolate the problem and here is what we found:
- host network traffic using node interface (ens5): all qdisc / tx queues are used
- host network traffic using pod interface (ens6): all qdisc / tx queues are used (to achieve this we added an available pod IP to ens6 directly and forced outgoing traffic to use it for a test destination using `ip rule to X lookup 11`
- pod traffic, routed through ens6: only first qdisc / first queue used
- manually created netns, traffic routed through ens6: only first qdisc / first queue used (expected of course as we used the same routing strategy as cilium)
- manually created netns, traffic routed through ens5: only first qdisc / first queue used (we allocated an additional IP to the ens5 ENI and added the necessary routing)

We performed the last 2 tests again but this time we created the veth with numtxqueues=4 on a node with 2 tx queues and this solved the issue (all qdisc / tx queues were used)

It looks like increasing rx/tx queues on veth devices will solve the issue and this is a pretty simple fix. However, I'm really not certain this is right answer though because we also run on GCP (with ipam=kubernetes and secondary cidrs for pods) and pod traffic is using all qdisc and tx queues (I checked on a node with 16 tx queues / qdisc and traffic is balanced across all of them).

On both AWS and GCP, root qdisc is mq and we have has many child ones as tx queues (as expected) using fq_codel (the Ubuntu default)

### Cilium Version

1.8.7

### Kernel Version

5.11.0-1022-aws and Ubuntu 20.04.1 on AWS
5.11.0-1020-gcp and Ubuntu 20.04.1 on GCP

### Kubernetes Version

1.19.1

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by lbernail** [comment]
After looking a bit more into this I think that the problem comes from this kernel patch to veth_xmit: https://github.com/torvalds/linux/commit/edbea922025169c0e5cdca5ebf7bf5374cc5566c. It impacts all traffic transiting through veth devices but it's impossible to mitigate when using the ENA driver because of the ENA implementation of `select_queue`
The patch was added in 5.12 but has been backported to multiple kernels and I've verified that instances using these kernels are impacted:
- kernels `5.11.11+`
- Ubuntu 20.04.3 kernels `5.11.0-10xx-aws` (at least all the ones I've looked at)
- Ubuntu 20.04.2 kernels `5.8.0-10xx-aws` (at least all the ones I've looked at)

I have confirmed that these other setups are not impacted:
- AWS, Ubuntu 20.04.1 with kernel `5.4.0-10xx-aws` (they don't have the veth_xmit patch)
- AWS, host network traffic: traffic does go through the veth_xmit code

More details on what is happening:
- when a packet comes from a network namespace through a veth, it goes through `veth_xmit` which, because of this commit https://github.com/torvalds/linux/commit/edbea922025169c0e5cdca5ebf7bf5374cc5566c, now records the queue_mapping in the skb (here we have a single tx queue on the veth, so skb_get_queue_mapping==0 and skb_record_rx_queue sets skb->queue_mapping to 1: https://github.com/torvalds/linux/blob/v5.12/drivers/net/veth.c#L301-L306) 
- in the host netns (after routing) `netdev_core_pick_tx` uses the network driver `select_queue` function if implemented and otherwise falls back to `netdev_pick_tx` (kernel default algorithm to select the queue) https://github.com/torvalds/linux/blob/v5.11/net/core/dev.c#L4034-L4043
- `ena_select_queue` uses the queue_mapping stored in the skb (otherwise it calls `netdev_pick_tx`): https://github.com/torvalds/linux/blob/v5.11/drivers/net/ethernet/amazon/ena/ena_netdev.c#L3117-L3131. skb_rx_queue_recorded==1 because of what happened in `veth_xmit`, which means `select_queue` returns skb_get_rx_queue: 0. The comment in the code is interesting:
```
	/* we suspect that this is good for in--kernel network services that
	 * want to loop incoming skb rx to tx in normal user generated traffic,
	 * most probably we will not get to this
	 */
```
- All traffic originally from a network namespace through a veth now leaves the instance uses tx queue 0

I was able to verify this with bpftrace by adding kprobes to interesting functions in the datapath to inspect the content of `skb->queue_mapping` (SKBQM):
```
         kprobe:dev_queue_xmit: skb:  0xffff989d03e22ae0     dev: 15               eth0  SKBQM:  0
       kprobe:__dev_queue_xmit: skb:  0xffff989d03e22ae0     dev: 15               eth0  SKBQM:  0
    kprobe:netdev_core_pick_tx: skb:  0xffff989d03e22ae0     dev: 15               eth0  SKBQM:  0
    kprobe:dev_hard_start_xmit: skb:  0xffff989d03e22ae0     dev: 15               eth0  SKBQM:  0
              kprobe:veth_xmit: skb:  0xffff989d03e22ae0     dev: 15               eth0  SKBQM:  0
      kprobe:__dev_forward_skb: skb:  0xffff989d03e22ae0     dev: 15               eth0  SKBQM:  1
       kprobe:skb_scrub_packet: skb:  0xffff989d03e22ae0     dev: 15               eth0  SKBQM:  1
               kprobe:netif_rx: skb:  0xffff989d03e22ae0     dev: 16    lxc9dc9781b7ff1  SKBQM:  1
     kprobe:enqueue_to_backlog: skb:  0xffff989d03e22ae0     dev: 16    lxc9dc9781b7ff1  SKBQM:  1
    kprobe:__netif_receive_skb: skb:  0xffff989d03e22ae0     dev: 16    lxc9dc9781b7ff1  SKBQM:  1
                 kprobe:ip_rcv: skb:  0xffff989d03e22ae0     dev: 16    lxc9dc9781b7ff1  SKBQM:  1
          kprobe:ip_rcv_finish: skb:  0xffff989d03e22ae0     dev: 16    lxc9dc9781b7ff1  SKBQM:  1
   kprobe:ip_route_input_noref: skb:  0xffff989d03e22ae0     dev: 16    lxc9dc9781b7ff1  SKBQM:  1
    kprobe:ip_route_input_slow: skb:  0xffff989d03e22ae0     dev: 16    lxc9dc9781b7ff1  SKBQM:  1
             kprobe:ip_forward: skb:  0xffff989d03e22ae0     dev: 16    lxc9dc9781b7ff1  SKBQM:  1
         kprobe:dev_queue_xmit: skb:  0xffff989d03e22ae0     dev:  4               ens6  SKBQM:  1
       kprobe:__dev_queue_xmit: skb:  0xffff989d03e22ae0     dev:  4               ens6  SKBQM:  1
    kprobe:netdev_core_pick_tx: skb:  0xffff989d03e22ae0     dev:  4               ens6  SKBQM:  1
       kprobe:ena_select_queue: skb:  0xffff989d03e22ae0     dev:  4               ens6  SKBQM:  1
        kprobe:sch_direct_xmit: skb:  0xffff989d03e22ae0     dev:  4               ens6  SKBQM:  0
    kprobe:dev_hard_start_xmit: skb:  0xffff989d03e22ae0     dev:  4               ens6  SKBQM:  0
         kprobe:ena_start_xmit: skb:  0xffff989d03e22ae0     dev:  4               ens6  SKBQM:  0
```
The interesting transitions:
- after `netdev_core_pick_tx` SKBQM==0 (a single tx queue on the veth)
- after `veth_xmit` SKBQM==1 (queue_mapping recorded in the SKB)
- after `ena_select_queue`, SKBQM==0 (queue_mapping restored, other options to pick the queue ignored)

Solving this problem likely means changing the logic of `ena_select_queue` (or simply removing the function) but this is not enough because `netdev_pick_tx` itself uses a similar logic when XPS is disabled: https://github.com/torvalds/linux/blob/v5.11/net/core/dev.c#L4002-L4015
- if XPS is enabled, use `get_xps_queue`, which will pick the tx queue based on XPS settings
- otherwise, call `skb_tx_hash`
- `skb_tx_hash` will also use the recorded queue_mapping if it is set (and otherwise fall back to using a hash to pick the queue, which is the behavior we expect in most cases)

I could verify this on GCP where instances use the virtio_net driver that does not implement `select_queue` and directly uses `netdev_pick_tx`
- with XPS enabled (the default setting: CPU to tx queue mappings are configured by the google_guest_agent): all queues are used (as long as processes use multiple cores, and it is easy to steer traffic by binding a process to a specific core)
- with XPS disabled: we also only use tx queue 0

I'm interested in your thoughts on how we could address this. A few ideas:
- set numtxqueues on veth devices to match the number of tx queues on the physical device (only a mitigation of course but this we can achieve with Cilium)
- discuss with the ENA team how we can modify `ena_select_queue` to simply rely on `netdev_pick_tx`. This does not completely address the problem, but with such a patch we will be able to mitigate the issue by enabling XPS (which is currently ignored for traffic coming from a veth)
- discuss with the kernel network team how to improve this datapath. I have no context on the change to veth_xmit that triggers the problem, but we should probably be able to leverage multiple tx queues by default for traffic coming from a veth regardless of XPS

**Comment by wdauchy** [comment]
> set numtxqueues on veth devices to match the number of tx queues on the physical device (only a mitigation of course but this we can achieve with Cilium)
> discuss with the kernel network team how to improve this datapath. I have no context on the change to veth_xmit that triggers the problem, but we should probably be able to leverage multiple tx queues by default for traffic coming from a veth regardless of XPS

this should be easier with newer kernel as https://github.com/torvalds/linux/commit/9d3684c24a5232c2d7ea8f8a3e60fe235e6a9867 "veth: create by default nr_possible_cpus queues" (>= v5.14.x), even though those are not activated by default

**Comment by borkmann** [comment]
@lbernail thanks for the excellent bug report!

Quick question, does it work to utilise multiqueue on ena when you enable Cilium's bwm (https://cilium.io/blog/2020/11/10/cilium-19#bwmanager, https://docs.cilium.io/en/v1.11/gettingstarted/bandwidth-manager/).

I suspect on 1.11 when you only set `bandwidthManager=true`, it should work as well as a workaround (w/o using bwm policies and such, iow, just to have it enabled).

bwm sets queue_mapping internally to 0, so ena and other drivers will pick tx queue.

We can do it generically from Cilium BPF side in that case as quick fix.

**Comment by lbernail** [comment]
@borkmann That's very interesting. We currently run 1.8 and we are evaluating 1.10/1.11 but we should be able to test quite easily. I'll let you know if it works as expected

I had a quick look at the bwm code and if I understood correctly the queue_mapping is reset in to-netdev which is attached to the instance "physical interface" right? (I'm trying to see if we can easily backport this but it's not so obvious in our setup because we don't load anything in to-netdev). The problem is we can't do it in from-container because veth_xmit will happen after the tc ebpf hook.

What do you think about the veth_xmit patch triggering the issue? I'm not sure what would be the best way to discuss it with upstream

**Comment by borkmann** [comment]
> The problem is we can't do it in from-container because veth_xmit will happen after the tc ebpf hook.

Why do you think that is? I don't think so.

We should be able to cleanly fix it, e.g. I just crafted https://github.com/cilium/cilium/pull/18388 as an option.

Could you check if that addresses it on your side? Then you don't need bwm, and it works with per-ep routes.

> What do you think about the veth_xmit patch triggering the issue? I'm not sure what would be the best way to discuss it with upstream.

If you like I can propose a fix for upstream kernel as well.

**Comment by lbernail** [comment]
> > The problem is we can't do it in from-container because veth_xmit will happen after the tc ebpf hook.
> Why do you think that is? I don't think so.

Right yes, I was thinking the ebpf code was injected on clsact/egress on the veth interface in the pod netns but it's actually on the other peer veth in the host netns.

> We should be able to cleanly fix it, e.g. I just crafted #18388 as an option w/o kernel upgrade.
> Could you check if that addresses it on your side? Then you don't need bwm, and it works with per-ep routes.

Thanks! I'll check very soon and I'll let you know

> > What do you think about the veth_xmit patch triggering the issue? I'm not sure what would be the best way to discuss it with upstream.
> If you like I can propose a fix for upstream kernel as well.

It's would probably be interesting because I assume this patch impacts multiple other setups. More than happy to help even if I'm sure you have for more context than I do

**Comment by borkmann** [comment]
> > > The problem is we can't do it in from-container because veth_xmit will happen after the tc ebpf hook.
> > > Why do you think that is? I don't think so.
> 
> Right yes, I was thinking the ebpf code was injected on clsact/egress on the veth interface in the pod netns but it's actually on the other peer veth in the host netns.

Yes exactly.

> > We should be able to cleanly fix it, e.g. I just crafted #18388 as an option w/o kernel upgrade.
> > Could you check if that addresses it on your side? Then you don't need bwm, and it works with per-ep routes.
> 
> Thanks! I'll check very soon and I'll let you know

Awesome, thanks, looking forward!

> > > What do you think about the veth_xmit patch triggering the issue? I'm not sure what would be the best way to discuss it with upstream.
> > > If you like I can propose a fix for upstream kernel as well.
> 
> It's would probably be interesting because I assume this patch impacts multiple other setups. More than happy to help even if I'm sure you have for more context than I do

Sounds good, and yes, agree that it affects multiple other, non-BPF related setups, too. I'll keep you posted wrt netdev.

**Comment by borkmann** [comment]
(Kernel fix for discussion sent here: https://patchwork.kernel.org/project/netdevbpf/patch/ef4cf3168907944502c81d8bf45e24eea1061e47.1641427152.git.daniel@iogearbox.net/)

**Comment by lbernail** [comment]
@borkmann I added the patch you suggested to our 1.8 branch and it works great (the patch I tested is even simpler as we don't run Cilium on any kernel older than 5.1) and it worked great!

- I got twice the throughput on a m5.large instance using iperf3. This number does not mean much as these instances have burstable networking, but the throughput I got was still much higher than anything I got without the patch during my previous tests.
- m5.large have 2 tx queues and we now use both
```
sudo tc -s qdisc show dev ens6
qdisc mq 0: root
 Sent 15047984073 bytes 1703868 pkt (dropped 0, overlimits 0 requeues 212)
 backlog 0b 0p requeues 212
qdisc fq_codel 0: parent :2 limit 10240p flows 1024 quantum 9015 target 5.0ms interval 100.0ms memory_limit 32Mb ecn
 Sent 7826981030 bytes 883831 pkt (dropped 0, overlimits 0 requeues 100)
 backlog 0b 0p requeues 100
  maxpacket 63105 drop_overlimit 0 new_flow_count 37896 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 0: parent :1 limit 10240p flows 1024 quantum 9015 target 5.0ms interval 100.0ms memory_limit 32Mb ecn
 Sent 7221003043 bytes 820037 pkt (dropped 0, overlimits 0 requeues 112)
 backlog 0b 0p requeues 112
  maxpacket 63105 drop_overlimit 0 new_flow_count 26282 ecn_mark 0
  new_flows_len 0 old_flows_len 0
```
 - looking at the skb->queue_mapping field in the stack, we can see it was reset in bpf_lxc, as expected:
```
         kprobe:dev_queue_xmit: skb:  0xffff9fa3952c50e0 dev:  7                 eth0 #TXQ:1 SKBQM:      0
       kprobe:__dev_queue_xmit: skb:  0xffff9fa3952c50e0 dev:  7                 eth0 #TXQ:1 SKBQM:      0
    kprobe:netdev_core_pick_tx: skb:  0xffff9fa3952c50e0 dev:  7                 eth0 #TXQ:1 SKBQM:      0
    kprobe:dev_hard_start_xmit: skb:  0xffff9fa3952c50e0 dev:  7                 eth0 #TXQ:1 SKBQM:      0
              kprobe:veth_xmit: skb:  0xffff9fa3952c50e0 dev:  7                 eth0 #TXQ:1 SKBQM:      0
      kprobe:__dev_forward_skb: skb:  0xffff9fa3952c50e0 dev:  7                 eth0 #TXQ:1 SKBQM:      1
       kprobe:skb_scrub_packet: skb:  0xffff9fa3952c50e0 dev:  7                 eth0 #TXQ:1 SKBQM:      1
               kprobe:netif_rx: skb:  0xffff9fa3952c50e0 dev:  8      lxc64e6d1d83689 #TXQ:1 SKBQM:      1
     kprobe:enqueue_to_backlog: skb:  0xffff9fa3952c50e0 dev:  8      lxc64e6d1d83689 #TXQ:1 SKBQM:      1
    kprobe:__netif_receive_skb: skb:  0xffff9fa3952c50e0 dev:  8      lxc64e6d1d83689 #TXQ:1 SKBQM:      1
                 kprobe:ip_rcv: skb:  0xffff9fa3952c50e0 dev:  8      lxc64e6d1d83689 #TXQ:1 SKBQM:      0 <====
          kprobe:ip_rcv_finish: skb:  0xffff9fa3952c50e0 dev:  8      lxc64e6d1d83689 #TXQ:1 SKBQM:      0
   kprobe:ip_route_input_noref: skb:  0xffff9fa3952c50e0 dev:  8      lxc64e6d1d83689 #TXQ:1 SKBQM:      0
    kprobe:ip_route_input_slow: skb:  0xffff9fa3952c50e0 dev:  8      lxc64e6d1d83689 #TXQ:1 SKBQM:      0
             kprobe:ip_forward: skb:  0xffff9fa3952c50e0 dev:  8      lxc64e6d1d83689 #TXQ:1 SKBQM:      0
         kprobe:dev_queue_xmit: skb:  0xffff9fa3952c50e0 dev:  4                 ens6 #TXQ:2 SKBQM:      0
       kprobe:__dev_queue_xmit: skb:  0xffff9fa3952c50e0 dev:  4                 ens6 #TXQ:2 SKBQM:      0
    kprobe:netdev_core_pick_tx: skb:  0xffff9fa3952c50e0 dev:  4                 ens6 #TXQ:2 SKBQM:      0
       kprobe:ena_select_queue: skb:  0xffff9fa3952c50e0 dev:  4                 ens6 #TXQ:2 SKBQM:      0
         kprobe:netdev_pick_tx: skb:  0xffff9fa3952c50e0 dev:  4                 ens6 #TXQ:2 SKBQM:      0
        kprobe:sch_direct_xmit: skb:  0xffff9fa3952c50e0 dev:  4                 ens6 #TXQ:2 SKBQM:      0
    kprobe:dev_hard_start_xmit: skb:  0xffff9fa3952c50e0 dev:  4                 ens6 #TXQ:2 SKBQM:      0
         kprobe:ena_start_xmit: skb:  0xffff9fa3952c50e0 dev:  4                 ens6 #TXQ:2 SKBQM:      0
```

And your kernel patch has already been accepted!
 
Thanks again for your help

**Comment by borkmann** [comment]
Very cool! Cloud costs are now cut by a large margin I assume! :-)

Thanks again for this awesome bug report and for you help in testing, Laurent!

**Comment by dtaht** [comment]
I just watched the related presentation on this over here: https://www.youtube.com/watch?v=7Rat98r7B3U @lbernail - which was very good. As much as I am a fq_codel fan, the problem here was being unable to service the single queue fast enough to achieve the throughput possible on the link, via fq_codel, at the 5ms target for delay within the box.

However, I have two thoughts. Two iperfs is not a realistic workload, and I try to stress that with sch_fq, what you get is ever increasing in-application latencies *after you achieve full throughput*.

So I try to encourage folk to throw many more flows at the link, and measure the RTT observed by the streams, using tools such as flent's tcp_nup test.  At some point with cubic and fq_codel, it should start dropping packets appropriately to keep the delay under control at the full throughput achievable. Alternatively, BBR can be used with sch_fq or fq_codel which also observes delay, but has some trouble competing with itself.

Secondly, RFC3168 ECN could have been used here to get rid of the retransmits even with fq_codel triggering too often, and after, with the 8 queues active and filling the link.

What was the outcome here?

**Comment by lbernail** [comment]
@dtaht Thanks for your comment! Once we mitigated the kernel bug and were able to use all the queues which solved our issues. We mostly used iperf to validate which tx queues we were using but we observed much better behavior with our workloads. If I remember correctly we had two issues:
- latency was high (very likely because we were using a single queue in the NIC)
- throughput was lower than expected : I think the problem came from fq_codel drops. fq likely mitigated this at the cost of increased latency

This issue triggered several discussions. We considered moving to fq, which seems to be a potential good solution for server workloads but given we sometimes run many containers on a host we are likely taking advantage of fq_codel if some applications are using all the throughput
